apis/Google.Storage.V1/Google.Storage.V1/SignedUrlUtils.cs

```diff
@@ -0,0 +1,396 @@
+        /// See https://cloud.google.com/storage/docs/access-control/signed-urls for more information on Signed URLs.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">
+        /// The <paramref name="credentialFilePath"/> does not refer to a valid JSON service account key file.
+        /// </exception>
+        /// <returns>
+        /// The Signed URL which can be used to provide access to a bucket or object for a limited amount of time.
+        /// </returns>
+        public static string Create(
+            string bucket,
+            string objectName,
+            string credentialFilePath,
+            DateTimeOffset? expiration,
+            HttpMethod requestMethod = null,
+            Dictionary<string, IEnumerable<string>> requestHeaders = null,
+            Dictionary<string, IEnumerable<string>> contentHeaders = null)
+        {
+            StorageClientImpl.ValidateBucketName(bucket);
+            GaxPreconditions.CheckNotNull(credentialFilePath, nameof(credentialFilePath));
+
+            var credentials = ParseCredentials(credentialFilePath);
+            var timestamp = ((int?)((expiration - UnixEpoch)?.TotalSeconds))?.ToString();
+            var resourcePath = $"/{bucket}";
+            if (objectName != null)
+            {
+                resourcePath += $"/{objectName}";
+            }
+            var extensionHeaders = GatherExtensionHeaders(requestHeaders, contentHeaders);
+
+            var contentMD5 = GetFirstHeaderValue(contentHeaders, "Content-MD5");
+            var contentType = GetFirstHeaderValue(contentHeaders, "Content-Type");
+
+            var signatureLines = new List<string>
+            {
+                (requestMethod ?? HttpMethod.Get).ToString(),
+                contentMD5,
+                contentType,
+                timestamp
+            };
+            signatureLines.AddRange(extensionHeaders.Select(
+                header => $"{header.Key}:{string.Join(", ", header.Value)}"));
+            signatureLines.Add(resourcePath);
+
+            var signature = CreateSignature(string.Join("\n", signatureLines), credentials);
+
+            var queryParameters = new List<string> { $"GoogleAccessId={credentials.Id}" };
+            if (timestamp != null)
+            {
+                queryParameters.Add($"Expires={timestamp}");
+            }
+            queryParameters.Add($"Signature={signature}");
+            return $"{StorageHost}{resourcePath}?{string.Join("&", queryParameters)}";
+        }
+        
+        private static string Base64Encode(this byte[] value)
+        {
+            return Convert.ToBase64String(value);
+        }
+
+        private static string CreateSignature(string data, ServiceAccountCredential.Initializer initializer)
+        {
+            // TODO: This is taken from ServiceAccountCredential. Expose this somehow so we don't need to duplicate logic.
+            var hashAlg = SHA256.Create();
```

Comment:
Done (in a few other places as well with MD5 and RandomNumberGenerator). When I cleanup the TODO above and share things between this and ServiceAccountCredential, I'll clean that one up as well.